presto-main/src/test/java/io/prestosql/sql/planner/TestTableScanRedirectionWithPushdown.java

```diff
@@ -0,0 +1,350 @@
+        }
+    }
+
+    private LocalQueryRunner createLocalQueryRunner(
+            ApplyTableScanRedirect applyTableScanRedirect,
+            Optional<ApplyProjection> applyProjection,
+            Optional<ApplyFilter> applyFilter)
+    {
+        LocalQueryRunner queryRunner = LocalQueryRunner.create(MOCK_SESSION);
+        MockConnectorFactory.Builder builder = MockConnectorFactory.builder()
+                .withGetTableHandle((session, schemaTableName) -> new MockConnectorTableHandle(schemaTableName))
+                .withGetColumns(name -> {
+                    if (name.equals(sourceTable)) {
+                        return ImmutableList.of(
+                                new ColumnMetadata(sourceColumnNameA, INTEGER),
+                                new ColumnMetadata(sourceColumnNameB, INTEGER));
+                    }
+                    else if (name.equals(destinationTable)) {
+                        return ImmutableList.of(
+                                new ColumnMetadata(destinationColumnNameA, INTEGER),
+                                new ColumnMetadata(destinationColumnNameB, INTEGER));
+                    }
+                    throw new IllegalArgumentException();
+                })
+                .withApplyTableScanRedirect(applyTableScanRedirect);
+        applyProjection.ifPresent(builder::withApplyProjection);
+        applyFilter.ifPresent(builder::withApplyFilter);
+
+        queryRunner.createCatalog(MOCK_CATALOG, builder.build(), ImmutableMap.of());
+        return queryRunner;
+    }
+
+    private Optional<ProjectionApplicationResult<ConnectorTableHandle>> mockApplyProjection(
+            ConnectorSession session,
+            ConnectorTableHandle tableHandle,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        MockConnectorTableHandle handle = (MockConnectorTableHandle) tableHandle;
+
+        List<ColumnHandle> newColumns = assignments.values().stream()
+                .collect(toImmutableList());
+        if (handle.getColumns().isPresent() && newColumns.equals(handle.getColumns().get())) {
+            return Optional.empty();
+        }
+
+        return Optional.of(
+                new ProjectionApplicationResult<>(
+                        new MockConnectorTableHandle(handle.getTableName(), handle.getConstraint(), Optional.of(newColumns)),
+                        projections,
+                        assignments.entrySet().stream()
+                                .map(assignment -> new Assignment(
+                                        assignment.getKey(),
+                                        assignment.getValue(),
+                                        ((MockConnectorColumnHandle) assignment.getValue()).getType()))
+                                .collect(toImmutableList())));
+    }
+
+    private Optional<ConstraintApplicationResult<ConnectorTableHandle>> mockApplyFilter(
```

Comment:
remove? You can just use `getMockApplyFilter(...)` instead (it's almost identical)